engines/freescape/games/driller.cpp

```diff
@@ -0,0 +1,538 @@
+	// if (isAmiga())
+	//	_viewArea = Common::Rect(72, 66, 567, 269);
+	// else
+	_viewArea = Common::Rect(40, 16, 279, 116);
+	_playerHeightNumber = 1;
+	_playerHeights.push_back(16);
+	_playerHeights.push_back(48);
+	_playerHeights.push_back(80);
+	_playerHeights.push_back(112);
+
+	_playerHeight = _playerHeights[_playerHeightNumber];
+	_playerWidth = 12;
+	_playerDepth = 32;
+
+	_initialProveEnergy = 48;
+	_initialProveShield = 50;
+	_initialJetEnergy = 29;
+	_initialJetShield = 34;
+}
+
+void DrillerEngine::gotoArea(uint16 areaID, int entranceID) {
+	debugC(1, kFreescapeDebugMove, "Jumping to area: %d, entrance: %d", areaID, entranceID);
+	if (!_gameStateBits.contains(areaID))
+		_gameStateBits[areaID] = 0;
+
+	assert(_areaMap.contains(areaID));
+	_currentArea = _areaMap[areaID];
+	_currentArea->show();
+
+	_currentAreaMessages.clear();
+	if (_messagesList.size() > 2) {
+		if (_currentArea->gasPocketRadius > 0)
+			_currentAreaMessages.push_back(_messagesList[1]);
+		else
+			_currentAreaMessages.push_back(_messagesList[2]);
+
+		_currentAreaMessages.push_back(_currentArea->name);
+	}
+
+	if (entranceID > 0 || areaID == 127) {
+		traverseEntrance(entranceID);
+	} else if (entranceID == 0) {
+		Math::Vector3d diff = _lastPosition - _position;
+		// debug("dif: %f %f %f", diff.x(), diff.y(), diff.z());
+		//  diff should be used to determinate which entrance to use
+		int newPos = -1;
+		if (abs(diff.x()) < abs(diff.z())) {
+			if (diff.z() > 0)
+				newPos = 4000;
+			else
+				newPos = 100;
+			_position.setValue(2, newPos);
+		} else {
+			if (diff.x() > 0)
+				newPos = 4000;
+			else
+				newPos = 100;
+			_position.setValue(0, newPos);
+		}
+		assert(newPos != -1);
+	}
+	_lastPosition = _position;
+
+	debugC(1, kFreescapeDebugMove, "starting player position: %f, %f, %f", _position.x(), _position.y(), _position.z());
+	playSound(5, false);
+	// Ignore sky/ground fields
+	if (_currentArea->getAreaFlags() == 1)
+		_gfx->_keyColor = 0;
+	else
+		_gfx->_keyColor = 255;
+
+	if (isAmiga() || isAtariST())
+		swapPalette(areaID);
+}
+
+void DrillerEngine::loadGlobalObjects(Common::SeekableReadStream *file, int offset) {
+	assert(!_areaMap.contains(255));
+	ObjectMap *globalObjectsByID = new ObjectMap;
+	file->seek(offset);
+	for (int i = 0; i < 8; i++) {
+		Object *gobj = load8bitObject(file);
+		assert(gobj);
+		assert(!globalObjectsByID->contains(gobj->getObjectID()));
+		debugC(1, kFreescapeDebugParser, "Adding global object: %d", gobj->getObjectID());
+		(*globalObjectsByID)[gobj->getObjectID()] = gobj;
+	}
+
+	_areaMap[255] = new Area(255, 0, globalObjectsByID, nullptr);
+}
+
+void DrillerEngine::loadAssets() {
+	if (isDemo())
+		loadAssetsDemo();
+	else
+		loadAssetsFullGame();
+
+	_angleRotations.push_back(5.3);
+	_angleRotations.push_back(9.5);
+
+	// Start playing music, if any
+	playMusic("Matt Gray - The Best Of Reformation - 07 Driller Theme.ogg");
```

Comment:
This is not part of the original game. Only the C64 release has music (which will be supported, when that release is implemented), however, the same guy who did the original music [recently released a restored version of the theme in an album](https://soundcloud.com/mattgrayc64/driller-2015-remake-preview). This code allows to play the background music, if the user bought the original album (same as other restored music from scummvm). Should I rename this to "track1" (there is only one track, anyway)

Reply:
Understood. No need to rename the track, but `playMusic` must be compilable with Vorbis not compiled. E.g. guard the relevant calls with `#ifdef USE_VORBIS`

Reply:
When you buy the album, you get different formats, including Vorbis (e.g. mp3, flac, etc). Should I change this to load any of these known formats? (there is already code somewhere to do that, I won't reinvent the wheel).

Reply:
Yes, support other formats since it is the case. 

We have code that will require renaming to `track1.ogg`, that is `Engine::existExtractedCDAudioFiles(uint track)` and `g_system->getAudioCDManager()->play()` but I'd suggest supporting files with original names

Reply:
This was finally implemented using `openStreamFile` which test all the supported codecs directly using the scummvm API.